### ISSUE TITLE
Squash merges and version not bumped

### ISSUE THREAD

**Issue by ben-foster-cko** [adamralph/minver#213]
Hi Adam,

I'm looking at MinVer as a replacement for GitVersion. I created a test repo and installed the global MinVer tool to see how it would calculate versions. Here's my commit tree:

```
*   9c29626 (HEAD -> master, tag: 0.1.1) Merge branch 'develop'
|\  
| *   73c376d (develop) Merge branch 'feature/1' into develop
| |\  
|/ /  
| * a59ed5c (feature/1) Feature 1
|/  
| * bcf5419 (feature/3) Feature 3
|/  
| * 90d4876 (feature/2) Feature 2
|/  
* 7bcf786 (tag: 0.1.0) Initial commit
```

I noticed a few things:

- When branching off feature branches the version remains the same as `develop`. Am I right in thinking that the only way to deploy feature artefacts would be to tag them explicitly?
- When squash merging a feature branch back into develop it didn't bump the version of `develop` (it remained as 0.1.0)
- When merging develop into master and tagging, then switching back to develop, it didn't bump the version - I would expect this to be the same as master (0.1.1) - perhaps due to using annotated tags?

Below is the full commit history:

```
➜  MinVerDemo git:(master) git commit -am "Initial commit"
➜  MinVerDemo git:(master) git checkout -b develop
Switched to a new branch 'develop'
➜  MinVerDemo git:(develop) minver
MinVer: No commit found with a valid SemVer 2.0 version. Using default version 0.0.0-alpha.0.
MinVer: Using { Commit: 7bcf786, Tag: null, Version: 0.0.0-alpha.0, Height: 0 }.
MinVer: Calculated version 0.0.0-alpha.0.
0.0.0-alpha.0
➜  MinVerDemo git:(develop) git checkout master
Switched to branch 'master'
➜  MinVerDemo git:(master) git tag -a 0.1.0 -m "Version 0.1.0"
➜  MinVerDemo git:(master) minver
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 0 }.
MinVer: Calculated version 0.1.0.
0.1.0
➜  MinVerDemo git:(master) git checkout develop
Switched to branch 'develop'
➜  MinVerDemo git:(develop) git checkout develop
Already on 'develop'
➜  MinVerDemo git:(develop) minver              
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 0 }.
MinVer: Calculated version 0.1.0.
0.1.0
➜  MinVerDemo git:(develop) git checkout -b feature/1
Switched to a new branch 'feature/1'
➜  MinVerDemo git:(feature/1) git commit -am "Feature 1"
[feature/1 a59ed5c] Feature 1
 1 file changed, 1 insertion(+), 1 deletion(-)
➜  MinVerDemo git:(feature/1) minver
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 1 }.
MinVer: Calculated version 0.1.1-alpha.0.1.
0.1.1-alpha.0.1
➜  MinVerDemo git:(feature/1) git checkout develop
Switched to branch 'develop'
➜  MinVerDemo git:(develop) minver                    
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 0 }.
MinVer: Calculated version 0.1.0.
0.1.0
➜  MinVerDemo git:(develop) git checkout -b feature/2
Switched to a new branch 'feature/2'
➜  MinVerDemo git:(feature/2) git commit -am "Feature 2"
[feature/2 90d4876] Feature 2
 1 file changed, 1 insertion(+), 1 deletion(-)
➜  MinVerDemo git:(feature/2) minver
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 1 }.
MinVer: Calculated version 0.1.1-alpha.0.1.
0.1.1-alpha.0.1
➜  MinVerDemo git:(feature/2) git checkout develop
Switched to branch 'develop'
➜  MinVerDemo git:(develop) minver
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 0 }.
MinVer: Calculated version 0.1.0.
0.1.0
➜  MinVerDemo git:(develop) git merge --squash feature/2
Updating 7bcf786..90d4876
Fast-forward
Squash commit -- not updating HEAD
 test.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
➜  MinVerDemo git:(develop) ✗ minver                      
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 0 }.
MinVer: Calculated version 0.1.0.
0.1.0
➜  MinVerDemo git:(develop) ✗ minver                      
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 0 }.
MinVer: Calculated version 0.1.0.
0.1.0
➜  MinVerDemo git:(develop) ✗ git checkout -b feature/3
M	test.txt
Switched to a new branch 'feature/3'
➜  MinVerDemo git:(feature/3) ✗ git commit -am "Feature 3"
[feature/3 bcf5419] Feature 3
 1 file changed, 1 insertion(+), 1 deletion(-)
➜  MinVerDemo git:(feature/3) minver
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 1 }.
MinVer: Calculated version 0.1.1-alpha.0.1.
0.1.1-alpha.0.1
➜  MinVerDemo git:(feature/3) git checkout develop
Switched to branch 'develop'
➜  MinVerDemo git:(develop) git checkout feature/1
Switched to branch 'feature/1'
➜  MinVerDemo git:(feature/1) minver                
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 1 }.
MinVer: Calculated version 0.1.1-alpha.0.1.
0.1.1-alpha.0.1
➜  MinVerDemo git:(feature/1) git checkout develop
Switched to branch 'develop'
➜  MinVerDemo git:(develop) git merge feature/1 --no-ff
Merge made by the 'recursive' strategy.
 test.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
➜  MinVerDemo git:(develop) minver
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 1 }.
MinVer: Calculated version 0.1.1-alpha.0.1.
0.1.1-alpha.0.1
➜  MinVerDemo git:(develop) git merge feature/1 --no-ff
➜  MinVerDemo git:(develop) git checkout feature/3
Switched to branch 'feature/3'
➜  MinVerDemo git:(feature/3) minver
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 1 }.
MinVer: Calculated version 0.1.1-alpha.0.1.
0.1.1-alpha.0.1
➜  MinVerDemo git:(feature/3) git checkout master
Switched to branch 'master'
➜  MinVerDemo git:(master) git merge develop --no-ff
Merge made by the 'recursive' strategy.
 test.txt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
➜  MinVerDemo git:(master) git tag -a 0.1.1 -m "Version 0.1.1"
➜  MinVerDemo git:(master) minver
MinVer: Using { Commit: 9c29626, Tag: '0.1.1', Version: 0.1.1, Height: 0 }.
MinVer: Calculated version 0.1.1.
0.1.1
➜  MinVerDemo git:(master) git checkout develop
Switched to branch 'develop'
➜  MinVerDemo git:(develop) minver
MinVer: Using { Commit: 7bcf786, Tag: '0.1.0', Version: 0.1.0, Height: 1 }.
MinVer: Calculated version 0.1.1-alpha.0.1.
0.1.1-alpha.0.1
```

**Comment by adamralph** [comment]
Hey @ben-foster-cko, thank for asking these questions.

The main thing to bear in mind is that MinVer performs no logic at all based on branch names. In fact, there is nothing in MinVer which even attempts to read which branches the current commit is contained in. The version is purely based on the last tag found in the history and the height of the current commit from that tag. There's more on this in the [README](https://github.com/adamralph/minver#how-it-works).

>     * When branching off feature branches the version remains the same as `develop`. Am I right in thinking that the only way to deploy feature artefacts would be to tag them explicitly?

Feature branches will indeed be versioned in the same way as develop, only the height will change as the current commit gets further away from the last tag. If you want a specific version to be built from your feature branch then yes, you can tag the commit. Or you can add some special logic yourself based on the name of the current branch using a similar technique to those described in the README for [performing custom logic](https://github.com/adamralph/minver#can-i-use-the-version-calculated-by-minver-for-other-purposes).

>     * When squash merging a feature branch back into develop it didn't bump the version of `develop` (it remained as 0.1.0)

Correct, for the reasons described above.

>     * When merging develop into master and tagging, then switching back to develop, it didn't bump the version - I would expect this to be the same as master (0.1.1) - perhaps due to using annotated tags?

Correct, for the reasons described above. Annotated tags should be fine, although I don't think I've tested them explicitly. I should probably add something into the tests to make sure they are handled correctly.

---

In short, don't expect MinVer to work the same way as GitVersion. When I started prototyping, I did attempt to introduce logic around branching, but it very quickly turned into a can of worms. MinVer is purely tag-based versioning, and branches are irrelevant.
